warp10/src/main/java/io/warp10/script/ext/http/HTTP.java

```diff
@@ -0,0 +1,437 @@
+        conn.setDoOutput(bodyB.length > 0);
+        if (bodyB.length > 0) {
+          try (OutputStream os = conn.getOutputStream()) {
+            os.write(bodyB);
+          }
+        }
+
+      } else {
+        throw new WarpScriptException(getName() + " expects the body of the request to be a STRING or BYTES object.");
+      }
+
+      //
+      // Form response
+      //
+
+      res.put(STATUS_CODE, conn.getResponseCode());
+      Map<String, List<String>> hdrs = conn.getHeaderFields();
+
+      if (hdrs.containsKey(null)) {
+        List<String> statusMsg = hdrs.get(null);
+        if (statusMsg.size() > 0) {
+          res.put(STATUS_MESSAGE, statusMsg.get(0));
+        } else {
+          res.put(STATUS_MESSAGE, "");
+        }
+      } else {
+        res.put(STATUS_MESSAGE, "");
+      }
+
+      //
+      // Make the headers map modifiable
+      //
+
+      hdrs = new HashMap<String, List<String>>(hdrs);
+      hdrs.remove(null);
+
+      res.put(RESPONSE_HEADERS, hdrs);
+
+      //
+      // Read response
+      //
+
+      InputStream in = null;
+
+      // When there is an error (response code is 404 for instance), body is in the error stream.
+      try {
+        in = conn.getInputStream();
+      } catch (IOException ioe) {
+        in = conn.getErrorStream();
+      }
+
+      if (chunkSize <= 0) {
+        byte[] buf = new byte[8192];
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+        while (true) {
+          int len = in.read(buf);
+          if (len < 0) {
+            break;
+          }
+
+          if (downloadSize.get() + baos.size() + len > maxsize) {
+            throw new WarpScriptException(getName() + " would exceed maximum size of content which can be retrieved via this function (" + maxsize + " bytes)");
+          }
+
+          baos.write(buf, 0, len);
+        }
+
+        downloadSize.addAndGet(baos.size());
+        res.put(CONTENT, baos.toByteArray());
+
+      } else {
+        byte[] buf = new byte[chunkSize.intValue()];
+        Map<String, Object> chunkRes = new HashMap<>(res);
+
+        int chunkNumber = 0;
+        while (true) {
+          chunkNumber++;
+
+          int len = in.read(buf);
+          if (len < 0) {
+            break;
+          }
+
+          if (downloadSize.addAndGet(len) > maxsize) {
+            throw new WarpScriptException(getName() + " would exceed maximum size of content which can be retrieved via this function (" + maxsize + " bytes)");
+          }
+
+          if (len == chunkSize) {
+            chunkRes.put(CONTENT, buf);
+          } else {
+            byte[] buf2 = new byte[len];
+            for (int i = 0; i < buf2.length; i++) {
+              buf2[i] = buf[i];
+            }
+            chunkRes.put(CONTENT, buf2);
+          }
+          chunkRes.put(CHUNK_NUMBER, new Long(chunkNumber));
+          stack.push(chunkRes);
+          if (null != chunkMacro) {
+            stack.exec(chunkMacro);
+          }
```

Comment:
If you make the macro optional, the content Map must be a copy, like discussed above.